### PR TITLE
java CDK: ContainerFactory properly supports unshared testcontainers

### PR DESCRIPTION
**Pull request by postamar**
Originally, the db-source's ContainerFactory test fixture was built under the assumption that testcontainers could always be shared by multiple tests. In practice it turns out that that's not always the case, especially for CDC tests where the CDC state is at the RDBMS level. For instance, in mysql, there's only one binlog.

Recently we struggled to debug an issue because an unshared testcontainer didn't forward its logs to the logging back-end. This is done by default for shared containers, but was forgotten for an unshared container in a CDC test. 

Rather than peppering these tests with the suitable logging incantations, this PR adds an `unshared` method to `ContainerFactory` which is like the existing `shared` except that, well, the container is not shared. This way testcontainers are provisioned in the same way regardless of whether they're shared or not.

This PR can be reviewed commit-by-commit.

### CODE REVIEW

File: airbyte-cdk/java/airbyte-cdk/airbyte-commons/src/main/java/io/airbyte/commons/logging/LoggingHelper.java [hunk 1]
```diff
@@ -22,6 +22,7 @@ public enum Color {
     YELLOW_BACKGROUND("\u001b[43m"), // destination
```
Comment (stephane-airbyte) [inline airbyte-cdk/java/airbyte-cdk/airbyte-commons/src/main/java/io/airbyte/commons/logging/LoggingHelper.java:22]: <i class='graphite__hidden'>[Re: line 13]<br/><br/></i>

I had no idea this existed. <span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/34865?utm_source=unchanged-line-comment">Graphite</a>.</span>

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java [hunk 2]
```diff
@@ -39,80 +29,14 @@ public interface ContainerFactory<C extends JdbcDatabaseContainer<?>> {
...
-              method.getName(), factory.getClass().getName(), imageName);
-          method.invoke(factory, sharedContainer);
-        }
-        sharedContainer.start();
-      } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
-        throw new RuntimeException(e);
-      }
-    }
-
+  default C unshared(String imageName, String... methods) {
```
Comment (stephane-airbyte) [inline airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java:38]: exclusive?

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactoryWrapper.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.JdbcDatabaseContainer;
+import org.testcontainers.containers.output.Slf4jLogConsumer;
+import org.testcontainers.utility.DockerImageName;
+
+/**
+ * This class wraps a specific shared testcontainer instance, which is created exactly once.
+ */
+class ContainerFactoryWrapper<C extends JdbcDatabaseContainer<?>> {
```
Comment (stephane-airbyte) [inline airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactoryWrapper.java:26]: does it need to be a JdbcDatabaseContainer? Or can we use GenericContainer here?

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactoryWrapper.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+  static private final Logger LOGGER = LoggerFactory.getLogger(ContainerFactoryWrapper.class);
+  static private final ConcurrentHashMap<String, ContainerFactoryWrapper<?>> SHARED_SINGLETONS = new ConcurrentHashMap<>();
+
+  static public final MdcScope.Builder TESTCONTAINER_LOG_MDC_BUILDER = new MdcScope.Builder()
+      .setLogPrefix("testcontainer")
+      .setPrefixColor(LoggingHelper.Color.RED_BACKGROUND);
+
+  @SuppressWarnings("unchecked")
+  static <C extends JdbcDatabaseContainer<?>> C getOrCreateShared(ContainerFactory<C> factory, String imageName, String... methods) {
```
Comment (stephane-airbyte) [inline airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactoryWrapper.java:36]: That's super weird to have a different C here than in the class definition.

Comment (postamar) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactoryWrapper.java:36]: It's static

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactoryWrapper.java (hunk 3 above)
Comment (postamar) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactoryWrapper.java:26]: Indeed we can.

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java (hunk 2 above)
Comment (postamar) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java:38]: Let's go with `exclusive`.

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactoryWrapper.java (hunk 4 above)
Comment (stephane-airbyte) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactoryWrapper.java:36]: I feel like all the methods here should be static, and the main class shouldn't be parameterized, no?

**Comment by postamar** [comment]
@stephane-airbyte I addressed your comments. I noticed the CI still fails so I'll wait for https://github.com/airbytehq/airbyte/pull/34182 to merge before anything else.

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactoryWrapper.java (hunk 4 above)
Comment (postamar) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactoryWrapper.java]: You're right, it's better when the class isn't parameterized.

**Comment by postamar** [comment]
closed in favour of https://github.com/airbytehq/airbyte/pull/34892

**Pull request closed without merging**